Userland/Libraries/LibCompress/Brotli.cpp

```diff
@@ -0,0 +1,905 @@
+/*
+ * Copyright (c) 2022, Michiel Visser <opensource@webmichiel.nl>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/BinarySearch.h>
+#include <AK/QuickSort.h>
+#include <LibCompress/Brotli.h>
+#include <LibCompress/BrotliDictionary.h>
+
+namespace Compress {
+
+ErrorOr<size_t> BrotliDecompressionStream::CanonicalCode::read_symbol(LittleEndianInputBitStream& input_stream)
+{
+    size_t code_bits = 1;
+
+    while (code_bits < (1 << 16)) {
+        // FIXME: This is very inefficient and could greatly be improved by implementing this
+        //        algorithm: https://www.hanshq.net/zip.html#huffdec
+        size_t index;
+        if (binary_search(m_symbol_codes.span(), code_bits, &index))
+            return m_symbol_values[index];
+
+        code_bits = (code_bits << 1) | TRY(input_stream.read_bit());
+    }
+
+    return Error::from_string_literal("no matching code found");
+}
+
+BrotliDecompressionStream::BrotliDecompressionStream(Stream& stream)
+    : m_input_stream(stream)
+{
+}
+
+ErrorOr<size_t> BrotliDecompressionStream::read_window_length()
+{
+    if (TRY(m_input_stream.read_bit())) {
+        switch (TRY(m_input_stream.read_bits(3))) {
+        case 0: {
+            switch (TRY(m_input_stream.read_bits(3))) {
+            case 0:
+                return 17;
+            case 1:
+                VERIFY_NOT_REACHED();
```

Comment:
This can be reached with this test file:
[crash-5fee003afd41dd02b9a21df2ae7f9749c57f7404.txt](https://github.com/SerenityOS/serenity/files/8421858/crash-5fee003afd41dd02b9a21df2ae7f9749c57f7404.txt)